### PR TITLE
Add ES module build

### PR DESCRIPTION
**Pull request by lewisl9029**
**Describe the pull request**

This PR adds a ES module entry point to the library at /esm/index.js in addition to the existing CommonJS entry point at /lib/index.js.

**Why**

Most new frontend tooling like esbuild, Vite, and most package CDNs these days are esm-first, i.e. CJS modules require extra work and helper functions to get converted to esm. Exposing an esm entry point saves some work for these tools and reduces code size due to removal of CJS conversion helpers.

**How**

I added a new .babelrc-esm config file that configures babel to output esm instead of cjs, and a `build:esm` script that makes use of the config to output to a /esm dir, and a new `"modules"` entry point in package.json to point to `/esm/index.js` for esm consumers.

**Screenshots**

N/A

As far as I can tell, this changes nothing about the existing CJS outputs and workflow, and only adds the ESM entry point on top. However, there is still a chance of breakage due to the fact that many tools resolve to the `module` entry point over `main` by default, so they'll automatically switch over to this new esm build once this is released. So if we wanted to be extra cautious, it may be worth releasing as a new major and noting this as a potential breaking change.

I've published a version with the changes in this PR [here](https://www.npmjs.com/package/@lewisl9029/intro.js-react), and tested with my own projects. So far they seem to work identically as before.

Happy to iterate on this until you're happy with it. Cheers!

### CODE REVIEW

**Comment by HiDeoo** [comment]
Thanks a lot for your pull request, really great and thorough work. I'm totally on board with the changes and agree with your reasoning.

- Thanks for typing out this in-depth description, covers everything I could think of.
- I like the approach of not adding new dependencies or new tooling, but rather using the existing ones in the project.
- The published version over `@lewisl9029/intro.js-react` is a great idea, allowed me to quickly test it in various demo sandboxes that I have and it worked great.

> So if we wanted to be extra cautious, it may be worth releasing as a new major and noting this as a potential breaking change.

I totally agree with that statement, specially considering the state of the ecosystem and how often we see packages having issues with the whole CJS/ESM thing. I think it's better to be safe than sorry.

The only thing I'm not 100% sure without further investigation are the `package.json` modifications and if we should also provides an `pkg.exports` field or not.

What I propose is the following:

- I still have various projects using `intro.js-react` and I'd like to test this new version in them before releasing it. I don't think from my early tests that there will be any issues, but I'd like to be sure.
- I would like to review the [report from publint.dev](https://publint.dev/@lewisl9029/intro.js-react@0.7.1-1) and see if some recommendations should be applied or not.

I'll try to do all of this as soon as I can.

If you have any thoughts or remarks, please let me know.

**Comment by lewisl9029** [comment]
Ah I didn't know about publint, those recommendations look great!

I think if we're going to do a major, we might as well do a few more potentially breaking changes so we don't have to do another one later down the line: 

- Add the exports field as it suggested and only expose the main index.js and maybe Step and Hint modules as the public API.
- Add `type: module` to package json and rename the CJS modules to .cjs instead of renaming the esm to .mjs, since those will be used more often in frontend projects
- Maybe remove the src dir from the files list to slim down the package size.

Happy to update the PR with those changes if you're onboard! :)

**Comment by lewisl9029** [comment]
Oh and it could be worth considering passing an explicit browserlist to preset-env to minimize the need for polyfills. If a user needs support for older browsers, they could always configure their bundlers to transpile the package themselves.

**Comment by HiDeoo** [comment]
> Add the exports field as it suggested and only expose the main index.js

Definitely agree.

> Add `type: module` to package json and rename the CJS modules to .cjs instead of renaming the esm to .mjs, since those will be used more often in frontend projects

I was about to say yes regarding this point and remembered that @markerikson talked a bit about ESM compat issues in Redux / RTK and also did a lot of research regarding this topic. 

I found [this recent PR](https://github.com/reduxjs/redux-toolkit/pull/3318) (and also [this comment](https://github.com/reduxjs/redux-toolkit/issues/3208#issuecomment-1444869358)) which I think could be a solid starting point considering all the tests and research that went into it, the webpack 4 issues, etc. Based on this, here is the configuration I suggest:

```json
"main": "dist/cjs/index.js",
"module": "dist/es/index.js",
"types": "index.d.ts",
"exports": {
  "./package.json": "package.json",
  ".": {
    "types": "index.d.ts",
    "import": "dist/es/index.mjs",
    "default": "dist/cjs/index.cjs"
  }
},
  ```

Some comments regarding this configuration:

- I suggest removing the existing `lib` folder and the newly introduced `esm` one and consolidate everything in `dist/cjs` & `dist/es` for consistency.
- `"module": "dist/es/index.js"` is explicitely a `.js` file (for webpack 4 support) which is just a copy of the `dist/es/index.mjs` file ([reference](https://github.com/reduxjs/redux/blob/a756061b65619a28f156c36377d8b93ae96a36e1/tsup.config.ts#L48-L49))
- In the `exports` field, the `"types": "index.d.ts",` has to be the first one as documented in https://www.typescriptlang.org/docs/handbook/esm-node.html#packagejson-exports-imports-and-self-referencing

> Maybe remove the src dir from the files list to slim down the package size.

Good catch. I have no idea why we have this in the first place. Looking at [the blame](https://github.com/HiDeoo/intro.js-react/commit/6186df7f735a199f1fbc67f6f0338d078e379190) this dates back to the initial release, not sure what I was thinking at the time. I think we can safely remove it.

> Oh and it could be worth considering passing an explicit browserlist to preset-env to minimize the need for polyfills.

Sounds like a good idea and I think this can be inlined directly in the babel config files, e.g. `"targets": "> 0.25%, not dead"`. Did you have any specific browserlist query in mind?

> Happy to update the PR with those changes if you're onboard! :)

First, do you have any thoughts or objections regarding my last suggested changes? Maybe I missed something? If not, I would be happy to review such changes if you're still willing to update the PR. Otherwise I'll get to them when I have some time.

Once the changes are in, I would still like to do a quick test in various projects that I have and after that I think we should be good to go and release a new version.

**Comment by markerikson** [comment]
@HiDeoo : hiya! Yeah, my own experience so far is that you should _not_ use `type: "module"` in `package.json`, and you _should_ use `.mjs` for ESM files.  (Except in the case of Webpack 4, in which case you should have a top-level `"module"` field that still has a `.js` extension.)  

Also, there's a really neat project called https://arethetypeswrong.github.io/ that will inspect your package and let you know if TS thinks the entry points are defined properly, and I have put together a prototype CLI tool that uses that library.  See https://github.com/arethetypeswrong/arethetypeswrong.github.io/issues/15 for discussion.

**Comment by markerikson** [comment]
@HiDeoo I know... I've been _begging_ for _years_ for someone who actually knows what they're doing to write such a comprehensive guide.

And no one has yet :(

I'm going to end up writing a blog post soon about what I've learned, but it's going to be "here's a bunch of stuff I've figured out by trial and error and I think half of that is still wrong", and definitely _not_ a comprehensive authoritative guide.

**Comment by HiDeoo** [comment]
> I'm going to end up writing a blog post soon

Would definitely be a great addition to your dev blog and also a solid, even if not authoritative, help to the community considering the state of the ecosystem.

**Comment by lewisl9029** [comment]
Wow, thanks for the discussion @HiDeoo and @markerikson, I never realized how deep this rabbit hole goes, learning so much haha.

I just pushed an update and published a new version for testing at https://www.npmjs.com/package/@lewisl9029/intro.js-react (version 0.7.1-2). Also ran it through arethetypeswrong and results are here: https://arethetypeswrong.github.io/?p=%40lewisl9029%2Fintro.js-react%400.7.1-2

Everything looks good except for this one point:

> Imports of "@lewisl9029/intro.js-react" under the node16 module resolution setting when the importing module is ESM (its extension is .mts or .mjs, or it has a .ts or .js extension and is in scope of a package.json that contains "type": "module") resolved to CJS types, but ESM implementations.

I believe this is because we're using the same `index.d.ts` declaration file for both the esm and cjs entry points, and due to the lack of `type: module`, it's being treated as a CJS module. When I first tried the package in my app, typechecking was broken with the following TypeScript error: 


seems to work fine, so maybe this is safe to ignore?

**Comment by lewisl9029** [comment]
Wow, thanks for the discussion @HiDeoo and @markerikson, I never realized how deep this rabbit hole goes, learning so much haha.

I just pushed an update and published a new version for testing at https://www.npmjs.com/package/@lewisl9029/intro.js-react (version 0.7.1-7). Also ran it through arethetypeswrong and results are here: https://arethetypeswrong.github.io/?p=%40lewisl9029%2Fintro.js-react%400.7.1-7

Everything looks good except for this one point:

> Imports of "@lewisl9029/intro.js-react" under the node16 module resolution setting when the importing module is ESM (its extension is .mts or .mjs, or it has a .ts or .js extension and is in scope of a package.json that contains "type": "module") resolved to CJS types, but ESM implementations.

I believe this is because we're using the same `index.d.ts` declaration file for both the esm and cjs entry points, and due to the lack of `type: module`, it's being treated as a CJS module. When I first tried the package in my app, typechecking was broken with the following TypeScript error: 

Screenshot 2023-04-19 at 5 44 50 PM

After I removed the `declare module` wrapper from index.d.ts though, typecheck seems to work fine, but the same warning is still there, so maybe this is safe to ignore?

A few other things worth noting:

- I consolidated the two babel configs into 1 .js config that branches based on env.
- I used `last 2 versions, not dead` as the browserlist query. Including `> 0.2%` seems to end up including IE11 in the list and results in a bunch of helpers being added. Without it the output is nice and clean. I think we can ask users to transpile them module themselves if they need IE11 support so we don't waste bytes for everybody else.
- Had to add [this plugin](https://www.npmjs.com/package/babel-plugin-add-import-extension) to rewrite imports to include extensions for the mjs version to work properly.

**Comment by markerikson** [comment]
@lewisl9029: yeah. That "FalseCJS" warning is what I'm getting for all the Redux alpha packages right now.

Per discussion with Andrew Branch from the TS team, the "Right" answer here is that you're supposed to somehow compile your types _twice_, once as "CJS" and once as "ESM", and have two different typedefs files.

I have not yet figured out a good way to do that in practice.  So, for now, I'm ignoring that warning.  (That of course means there's a good chance something will break :)  But there's only so many changes I can keep making to all this package configuration.)

**Comment by lewisl9029** [comment]
@markerikson I managed to find a setup that gets rid of the warning by adding another layer of conditions and copy pasting the index.d.ts to each subfolder with the appropriate .d.mts/.d.cts extensions:

```
  "exports": {
    "./package.json": "./package.json",
    ".": {
      "import": {
        "types": "./dist/esm/index.d.mts",
        "default": "./dist/esm/index.mjs"
      },
      "default": {
        "types": "./dist/cjs/index.d.cts",
        "default": "./dist/cjs/index.cjs"
      }
    }
  },
```

Here are the new results: https://arethetypeswrong.github.io/?p=%40lewisl9029%2Fintro.js-react%400.7.1-9

So it seems like the warning was actually only complaining about the extension, and the actual content of the declarations file be in esm form.

For the case of this module and its handwritten declarations, all I had to do was copy paste it with the right extension. If you're generating the declarations with tsc it might get a bit more complicated. But just copy pasting the esm declarations to the cjs dir with .d.cts could still be worth a try? 🤔

**Comment by lewisl9029** [comment]
Oh hey I just found this example in the handbook that seems to match this format pretty closely:

https://www.typescriptlang.org/docs/handbook/esm-node.html#packagejson-exports-imports-and-self-referencing

```
    "exports": {
        ".": {
            // Entry-point for `import "my-package"` in ESM
            "import": {
                // Where TypeScript will look.
                "types": "./types/esm/index.d.ts",
                // Where Node.js will look.
                "default": "./esm/index.js"
            },
            // Entry-point for `require("my-package") in CJS
            "require": {
                // Where TypeScript will look.
                "types": "./types/commonjs/index.d.cts",
                // Where Node.js will look.
                "default": "./commonjs/index.cjs"
            },
        }
    },
    // Fall-back for older versions of TypeScript
    "types": "./types/index.d.ts",
    // CJS fall-back for older versions of Node.js
    "main": "./commonjs/index.cjs"
```

Hopefully that means it should be reasonably well supported from the typescript side at least.

File: package.json [hunk 1]
```diff
@@ -2,12 +2,28 @@
   "name": "intro.js-react",
   "version": "0.7.1",
   "description": "Intro.js React Wrapper",
-  "main": "lib/index.js",
+  "main": "dist/lib/index.js",
```
Comment (HiDeoo) [inline package.json:5]:
```suggestion
  "main": "dist/cjs/index.js",
```

File: package.json [hunk 2]
```diff
@@ -2,12 +2,28 @@
...
+        "types": "./dist/cjs/index.d.cts",
+        "default": "./dist/cjs/index.cjs"
+      }
+    }
+  },
   "scripts": {
-    "prebuild": "rimraf lib/*",
-    "build": "babel --ignore '**/*.test.js' --out-dir lib src",
-    "build:watch": "npm run build -- --watch",
+    "prebuild": "rimraf lib/* && rimraf esm/*",
```
Comment (HiDeoo) [inline package.json:22]:
```suggestion
    "prebuild": "rimraf dist",
```

**Review by HiDeoo — CHANGES_REQUESTED** [review]
Fantastic investigation and changes @lewisl9029. 

I have suggested a few changes as part of the pull request review. On top of that, here are some additional comments:

- I tested locally and can confirm your findings using the `> 0.25%, not dead` browserlist query. I am a bit hesitant on using `last 2 versions, not dead` as-is and I think [`> 0.5%, last 2 versions, Firefox ESR, not dead`](https://browsersl.ist/#q=%3E+0.5%25%2C+last+2+versions%2C+Firefox+ESR%2C+not+dead) (which is basically the [`default`](https://browsersl.ist/#q=defaults) query) may be a safer bet. Right now, it produces the same output as `last 2 versions, not dead` if I am not wrong (tested using `git diff --no-prefix --no-index --binary dist dist-pr`) but may be more future-proof? Any thoughts?

- I think it is safe to remove the ambient module (`declare module`) from the `index.d.ts` file like you did in the PR.

- I think we should explicitely add `@babel/core` as a development dependency. Got 90+ warnings about this unmet peer dependency when running a fresh install, a bit annoying.

Not a lot more to say, the PR looks more and more complete to me. I am looking forward to seeing the last few changes implemented, do some more testing on existing projects and get this merged and released. Thanks again for your work on this!

**Comment by lewisl9029** [comment]
Thanks for catching those missed updates and the comments! I've updated the PR with all of the requested changes. Let me know if you find anything else.

**Comment by markerikson** [comment]
@lewisl9029 : 

> I managed to find a setup that gets rid of the warning by adding another layer of conditions and copy pasting the index.d.ts to each subfolder with the appropriate .d.mts/.d.cts extensions:

Yeah, that's theoretically an option here.  Andrew Branch from the TS team would likely tell you that there _could_ be subtle differences in the generated typedefs depending on whether you're treating it as an ESM module or a CJS module, thus the need to run `tsc` twice.  In practice, _hopefully_ that's not a concern and you can get away with copy-pasting the file.

**Review by HiDeoo — APPROVED** [review]

**Comment by HiDeoo** [comment]
> I've updated the PR with all of the requested changes. Let me know if you find anything else.

This looks great. I've slightly updated the Babel config for tests as Jest didn't like the consolidated new one as-is.

I'll start testing this new version in a few projects for a few days and I think we should be good. Do you mind updating `@lewisl9029/intro.js-react` with the recent changes? It would make testing a bit easier for me but if not possible, that is not an issue.

**Comment by lewisl9029** [comment]
Sure thing, just updated the package with the latest changes at version `0.7.1-12`. Let me know if you find anything!

**Comment by HiDeoo** [comment]
> Sure thing, just updated the package with the latest changes at version `0.7.1-12`. Let me know if you find anything!

Well, all my tests in a few projects went well, did not hit any issues, mostly drop-in replacements. I think we, and especially you, did everything to make the change as non-breaking as possible with all the informations we had.

I will merge this PR right now, and when I get home with a bit of free time, update the changelog with some warnings and publish a new major version `1.0.0`.

Thanks again for you amazing work and patience on this PR. I'll make sure to ping this thread when the new version is released.

**Comment by HiDeoo** [comment]
🎉  [`v1.0.0`](https://www.npmjs.com/package/intro.js-react/v/1.0.0) has been published to the npm registry. Thanks again for your work on this PR.

**Comment by lewisl9029** [comment]
Thank you for the review and great suggestions! I learned a ton. :)

**Pull request merged**